### PR TITLE
feat(ContentDiff): migrate all taxonomies

### PR DESCRIPTION
**Pull request by kariae**
## How to test
- Create a local site, and install a couple of plugins that use custom taxonomies ([Newspack Multibranded Site](https://github.com/automattic/newspack-multibranded-site), [Co-Authors Plus](https://wordpress.org/plugins/co-authors-plus/), ...).
- Add a couple posts, and link them to these custom taxonomies (setting post brands, and co-authors) as long as setting posts categories and tags.
1. Take a backup of the database (this will be our staging snapshot `staging.sql`).
2. Add more posts and custom taxonomies.
3. Rename all the database tables to use the prefix `live_` instead of `wp_`.
4. Import the backup from Step 1.
5. Create a logs folder somewhere in your computer `/tmp/test_cdiff/`.
6. Inside that folder create two files:
- An empty `content-diff__modified-ids.log` (unless if you updated some posts, then put their IDs in here in a CSV format)
- `content-diff__new-ids-csv.log` that contains the IDs of the new posts in a CSV format (p.s. if you're testing with CAP don't forget to add the co-authors entity IDs too because they are custom posts).
7. Run the migration command (Set the names of your custom taxonomies in the parameter `--custom-taxonomies-csv`):
`wp newspack-content-migrator content-diff-migrate-live-content --import-dir=/tmp/test_cdiff/ --live-table-prefix=live_ --custom-taxonomies-csv=author,brand`
8. Check that everything was migrated correctly.

---

- [x] confirmed that PHPCS has been run

### CODE REVIEW

**Review by naxoc — COMMENTED** [review]
I gave this a test locally and I think it worked. I would love more testing though. Maybe on actual publisher data on a clone somehow – I'm not sure I excercised enough of the code with my testing.

I left a small code comment

**Comment by eddiesshop** [comment]
This is super great Zak! Thanks for getting the ball rolling on this!

My only concern is that I think the `author` taxonomy is too much of a special case to migrate using this mechanism, because certain `author` taxonomy rows correspond directly to a `wp_posts` record (usually found via `wp_term_relationships`, although sometimes, somehow, that data gets lost for sites). I'm afraid that if we migrate these `author` taxonomy rows as such, without ensuring that the old corresponding `wp_posts` record exists first, and without updating the `wp_term_taxonomy.description` field after migrated, that we'll lose some important context around that data, and which would create misleading confusion if we ever needed to update/fix those records should any issue arise.

**Comment by kariae** [comment]
> without ensuring that the old corresponding wp_posts record exists first

The command that [the Knife tool ran before](https://github.com/Automattic/newspack-launch-army-knife/blob/29676328af8a612d22abdf45d80a8ac9fde3c598/inc/script_content_diff_from_jp_or_custom_archive.sh#L275-L280) this one is [content-diff-search-new-content-on-live](https://github.com/Automattic/newspack-custom-content-migrator/blob/23288c186b04f68d395f325bb37873663407f809/src/Command/General/ContentDiffMigrator.php#L380) which accepts `post-types` to migrate as a parameter, so there we need to set `guest-author` as a post type we need to migrate. Maybe we can add some logic in the knife tool to ensure that if we're migrating the `author` custom taxonomy, we've set `guest-author` on the previous command, but that's out of scope for this PR.

> without updating the wp_term_taxonomy.description field after migrated
Mhmm, actually we don't migrate the `wp_term_taxonomy.description` but I can add that.

**Comment by kariae** [comment]
@eddiesshop [4396f96](https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/4396f96b753ca8d0d3c2c9876b6a04cfffaef0eb) migrate the term description when it's added, and [d002704](https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/d0027046106de66e4762d3ef2ff274ad0933494c) update the description and the count when the term already exists. I did some tests and it looks fine, can you give it a shot, please?

File: src/Command/General/ContentDiffMigrator.php [hunk 1]
```diff
@@ -513,32 +532,56 @@ public function cmd_migrate_live_content( $args, $assoc_args ) {
...
+		if ( ! empty( $custom_taxonomies ) ) {
+			// Check if any of the custom taxonomies exist in the live DB.
+			$live_table_prefix_escaped = esc_sql( $live_table_prefix );
+			// phpcs:ignore -- table prefix string value was escaped.
+			$custom_taxonomies_live = $wpdb->get_col( "SELECT DISTINCT( taxonomy ) FROM {$live_table_prefix_escaped}term_taxonomy WHERE taxonomy NOT IN ('category', 'post_tag') ;" );
+			WP_CLI::log( sprintf( 'These unique custom taxonomies exist in live DB:%s', "\n- " . implode( "\n- ", $custom_taxonomies_live ) ) );
+
+			array_walk(
+				$custom_taxonomies,
+				function ( &$v ) use ( $custom_taxonomies_live, &$hierarchical_taxonomies_to_migrate, &$non_hierarchical_taxonomies_to_migrate ) {
```
[the remark this answers is not in the document]
Comment (kariae) [reply src/Command/General/ContentDiffMigrator.php]: That's part of the original code, I think a _foreach_ makes sense in this case. [56100ca](https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/56100ca8a96849c5bcebe8dfa45f540524f6a79a)

**Comment by eddiesshop** [comment]
I just finished testing this, and I think perhaps we should prevent the `author` taxonomy (and `guest-author` post-type) from being migrated. It is just too much of a special case.

From my testing, the updates in this PR will migrate the CPT `guest-author` and the custom taxonomy `author`, however it also does the following:
1. when the author taxonomies are migrated, for some reason their slugs are altered (cap-eddie-carrasco => eddie-carrasco)
2. None of the associations to stories the guest author is tied to are recreated.
3. The description column isn't updated during the migration, and this contains some useful metadata about which record the guest author is ultimately linked to. If left alone, as these GA's are used to assign to stories, they will get their description updated by the CAP plugin (see [here](https://github.com/Automattic/Co-Authors-Plus/blob/main/php/class-coauthors-plus.php#L1006) and [here](https://github.com/Automattic/Co-Authors-Plus/blob/main/php/class-coauthors-plus.php#L1625)). Then over time, what will happen is this data will have some info that is updated, and some that isn't. This might cause headaches if we need to debug something with this data should any problems arise, because we wouldn't be able to trust it completely (and I think no log record like we do for posts [old id => new id relationship]).

**Comment by eddiesshop** [comment]
While the previous test I did on this PR was focused on CAP, I re-did the testing focused on a custom taxonomy.

I created a custom taxonomy on the existing using `register_taxonomy('custom_taxonomy_for_testing', 'post')`. Then I inserted a new term using `wp_insert_term('test', 'custom_taxonomy_for_testing')`. From there, I pulled ~15 post ID randomly from the DB and inserted new relationship data for those posts, tying them to this  newly created taxonomy.

Finally, I ran the content diff, making sure to specify the flag: `--custom-taxonomies-csv=custom_taxonomy_for_testing`.

I got the following error message:
```
Error occurred while inserting custom_taxonomy_for_testing 'test' live_term_id=44815 at live_post_ID=1148717 :Invalid taxonomy.
```

I think perhaps there's a step missing where we need to use the `register_taxonomy` function on the DB where we're migrating this data to, if that taxonomy doesn't already exist?

Something else I noticed was that this error only appeared for 2 posts that were caught as being new (i.e. having to be content-diffed to the DB). I think this means that any posts that were already migrated, but which may or may not have had new taxonomies associated to them, aren't being picked up.

**Comment by iuravic** [comment]
Hi @eddiesshop, this now ready for a test run. Since it's an important piece of functionality, it would be great to once again give it an actual run with test data, please 🙏 

Also noting that I have not yet improved the fact that a post which has had a change in its taxonomy is not recognized as a modified post. Thanks for detecting that, it's an important find, I'd like to fix that in a separate PR right after we finish this one.

Changes I made:
- several fixes and changes in https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/f44a11d26da6efa9dbe089a9ebcf8f030b2a3bb9
- disallowed CAP's "author" taxonomy in https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/4e015fb96064ff52620072e8b32342ac86687a87

**Comment by iuravic** [comment]
Created https://github.com/Automattic/newspack-custom-content-migrator/pull/542 on top of this branch. Therefore, this branch needs to be merged first, and only then the branch from #542 `feat/migrate-all-wpusers`.

**Review by eddiesshop — CHANGES_REQUESTED** [review]
Finally got around to testing this! For the happy path, everything worked perfectly 👍 great job!

The only thing I would change would be adding a message similar to how you have it for the `author` taxonomy to explicitly prevent the `guest-author` CPT from being migrated.

I received no warning when I did this:
`wp newspack-content-migrator content-diff-search-new-content-on-live --export-dir=/var/www/html/test_content_diff_3 --live-table-prefix=live_wp_ --post-types-csv=post,page,attachment,guest-author`

**Comment by iuravic** [comment]
Thanks for the great catch, @eddiesshop , added prevention of `guest-author` CPT in [3fb1ade](https://github.com/Automattic/newspack-custom-content-migrator/pull/516/commits/3fb1adec17f8b1c09a6a52b14b11d91230052b4b).
Now merging!

**Pull request merged**